File: backend/auth/rbac.py
"""
Role-Based Access Control (RBAC) Service
Advanced permission management and role assignment
"""

from sqlalchemy.orm import Session
from typing import List, Optional, Dict, Any
from enum import Enum
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

class Permission(str, Enum):
    """Available permissions in the system"""
    # User management
    USERS_READ = "users:read"
    USERS_WRITE = "users:write"
    USERS_DELETE = "users:delete"
    USERS_MANAGE = "users:manage"
    
    # Role management
    ROLES_READ = "roles:read"
    ROLES_WRITE = "roles:write"
    ROLES_DELETE = "roles:delete"
    ROLES_MANAGE = "roles:manage"
    
    # Permission management
    PERMISSIONS_READ = "permissions:read"
    PERMISSIONS_WRITE = "permissions:write"
    PERMISSIONS_MANAGE = "permissions:manage"
    
    # Tenant management
    TENANTS_READ = "tenants:read"
    TENANTS_WRITE = "tenants:write"
    TENANTS_DELETE = "tenants:delete"
    TENANTS_MANAGE = "tenants:manage"
    
    # Property management
    PROPERTIES_READ = "properties:read"
    PROPERTIES_WRITE = "properties:write"
    PROPERTIES_DELETE = "properties:delete"
    PROPERTIES_MANAGE = "properties:manage"
    
    # Hotel configuration
    HOTEL_CONFIG_READ = "hotel_configuration:read"
    HOTEL_CONFIG_WRITE = "hotel_configuration:write"
    HOTEL_CONFIG_DELETE = "hotel_configuration:delete"
    HOTEL_CONFIG_MANAGE = "hotel_configuration:manage"
    
    # Booking management
    BOOKINGS_READ = "bookings:read"
    BOOKINGS_WRITE = "bookings:write"
    BOOKINGS_DELETE = "bookings:delete"
    BOOKINGS_MANAGE = "bookings:manage"
    
    # Financial management
    FINANCIAL_READ = "financial:read"
    FINANCIAL_WRITE = "financial:write"
    FINANCIAL_DELETE = "financial:delete"
    FINANCIAL_MANAGE = "financial:manage"
    
    # System management
    SETTINGS_READ = "settings:read"
    SETTINGS_WRITE = "settings:write"
    SETTINGS_MANAGE = "settings:manage"
    
    # Analytics
    ANALYTICS_READ = "analytics:read"
    ANALYTICS_WRITE = "analytics:write"
    ANALYTICS_MANAGE = "analytics:manage"

class Role(str, Enum):
    """Available roles in the system"""
    SUPER_ADMIN = "super_admin"
    ADMIN = "admin"
    MANAGER = "manager"
    STAFF = "staff"
    GUEST = "guest"

class UserRole(str, Enum):
    """Available user roles in the system"""
    SUPER_ADMIN = "super_admin"
    ADMIN = "admin"
    MANAGER = "manager"
    STAFF = "staff"
    GUEST = "guest"

class PermissionScope(str, Enum):
    """Permission scopes for fine-grained access control"""
    GLOBAL = "global"      # System-wide access
    TENANT = "tenant"      # Tenant-specific access
    PROPERTY = "property"  # Property-specific access
    USER = "user"          # User-specific access

class RBACService:
    """
    Role-Based Access Control Service
    Manages user roles, permissions, and access control
    """
    
    def __init__(self, db: Session):
        self.db = db
    
    def assign_role(self, user_id: str, role: UserRole, assigned_by: str) -> bool:
        """
        Assign a role to a user
        """
        try:
            from models.user import User
            from models.user_role import UserRoleAssignment
            
            user = self.db.query(User).filter(User.id == user_id).first()
            if not user:
                logger.error(f"User {user_id} not found for role assignment")
                return False
            
            # Check if user can assign this role
            assigner = self.db.query(User).filter(User.id == assigned_by).first()
            if not self._can_assign_role(assigner, role):
                logger.error(f"User {assigned_by} cannot assign role {role}")
                return False
            
            # Create or update role assignment
            existing_assignment = self.db.query(UserRoleAssignment).filter(
                UserRoleAssignment.user_id == user_id
            ).first()
            
            if existing_assignment:
                existing_assignment.role = role.value
                existing_assignment.assigned_by = assigned_by
                existing_assignment.assigned_at = datetime.utcnow()
            else:
                role_assignment = UserRoleAssignment(
                    user_id=user_id,
                    role=role.value,
                    assigned_by=assigned_by,
                    assigned_at=datetime.utcnow()
                )
                self.db.add(role_assignment)
            
            # Update user role
            user.role = role.value
            user.updated_at = datetime.utcnow()
            
            self.db.commit()
            logger.info(f"Role {role} assigned to user {user_id} by {assigned_by}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to assign role: {str(e)}")
            self.db.rollback()
            return False
    
    def revoke_role(self, user_id: str, revoked_by: str) -> bool:
        """
        Revoke user's role (set to guest)
        """
        try:
            from models.user import User
            from models.user_role import UserRoleAssignment
            
            user = self.db.query(User).filter(User.id == user_id).first()
            if not user:
                logger.error(f"User {user_id} not found for role revocation")
                return False
            
            # Check if user can revoke this role
            revoker = self.db.query(User).filter(User.id == revoked_by).first()
            if not self._can_revoke_role(revoker, user):
                logger.error(f"User {revoked_by} cannot revoke role from user {user_id}")
                return False
            
            # Update role assignment
            role_assignment = self.db.query(UserRoleAssignment).filter(
                UserRoleAssignment.user_id == user_id
            ).first()
            
            if role_assignment:
                role_assignment.role = UserRole.GUEST.value
                role_assignment.assigned_by = revoked_by
                role_assignment.assigned_at = datetime.utcnow()
            
            # Update user role
            user.role = UserRole.GUEST.value
            user.updated_at = datetime.utcnow()
            
            self.db.commit()
            logger.info(f"Role revoked from user {user_id} by {revoked_by}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to revoke role: {str(e)}")
            self.db.rollback()
            return False
    
    def get_user_permissions(self, user_id: str) -> List[str]:
        """
        Get all permissions for a user based on their role
        """
        try:
            from models.user import User
            from .permissions import PERMISSIONS_MAP
            
            user = self.db.query(User).filter(User.id == user_id).first()
            if not user:
                return []
            
            return PERMISSIONS_MAP.get(user.role, [])
            
        except Exception as e:
            logger.error(f"Failed to get user permissions: {str(e)}")
            return []
    
    def check_permission(self, user_id: str, permission: str, resource_id: Optional[str] = None) -> bool:
        """
        Check if user has specific permission, optionally for a specific resource
        """
        try:
            from models.user import User
            from .permissions import has_permission
            
            user = self.db.query(User).filter(User.id == user_id).first()
            if not user:
                return False
            
            # Check basic permission
            if not has_permission(user, permission):
                return False
            
            # Check resource-specific access if resource_id provided
            if resource_id:
                return self._check_resource_access(user, permission, resource_id)
            
            return True
            
        except Exception as e:
            logger.error(f"Failed to check permission: {str(e)}")
            return False
    
    def get_users_by_role(self, role: UserRole, tenant_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get all users with specific role, optionally filtered by tenant
        """
        try:
            from models.user import User
            
            query = self.db.query(User).filter(User.role == role.value)
            
            if tenant_id:
                query = query.filter(User.tenant_id == tenant_id)
            
            users = query.all()
            
            return [
                {
                    "id": user.id,
                    "email": user.email,
                    "full_name": user.full_name,
                    "role": user.role,
                    "is_active": user.is_active,
                    "created_at": user.created_at,
                    "last_login": user.last_login
                }
                for user in users
            ]
            
        except Exception as e:
            logger.error(f"Failed to get users by role: {str(e)}")
            return []
    
    def get_role_statistics(self, tenant_id: Optional[str] = None) -> Dict[str, Any]:
        """
        Get statistics about role distribution
        """
        try:
            from models.user import User
            from sqlalchemy import func
            
            query = self.db.query(User.role, func.count(User.id).label('count'))
            
            if tenant_id:
                query = query.filter(User.tenant_id == tenant_id)
            
            results = query.group_by(User.role).all()
            
            stats = {}
            total_users = 0
            
            for role, count in results:
                stats[role] = count
                total_users += count
            
            stats['total'] = total_users
            
            # Calculate percentages
            for role in stats:
                if role != 'total' and total_users > 0:
                    stats[f"{role}_percentage"] = round((stats[role] / total_users) * 100, 2)
            
            return stats
            
        except Exception as e:
            logger.error(f"Failed to get role statistics: {str(e)}")
            return {}
    
    def _can_assign_role(self, assigner, target_role: UserRole) -> bool:
        """
        Check if user can assign the target role
        """
        if not assigner:
            return False
        
        from .permissions import get_role_level, can_manage_role
        return can_manage_role(assigner.role, target_role.value)
    
    def _can_revoke_role(self, revoker, target_user) -> bool:
        """
        Check if user can revoke role from target user
        """
        if not revoker or not target_user:
            return False
        
        from .permissions import get_role_level, can_manage_role
        return can_manage_role(revoker.role, target_user.role)
    
    def _check_resource_access(self, user, permission: str, resource_id: str) -> bool:
        """
        Check if user has access to specific resource
        """
        # Super admins can access any resource
        if user.role == "super_admin":
            return True
        
        # Check tenant access
        if permission.startswith("tenants:") or permission.startswith("onboarding:"):
            return user.tenant_id == resource_id
        
        # Check property access
        if permission.startswith("properties:") or permission.startswith("rooms:") or permission.startswith("hotel_configuration:"):
            from models.hospitality_property import HospitalityProperty
            property = self.db.query(HospitalityProperty).filter(
                HospitalityProperty.id == resource_id
            ).first()
            return property and property.tenant_id == user.tenant_id
        
        # Check user access
        if permission.startswith("users:") or permission.startswith("profile:"):
            return user.id == resource_id or user.role in ["admin", "super_admin"]
        
        # Default to True for other permissions
        return True
    
    def create_custom_role(self, role_name: str, permissions: List[str], created_by: str) -> bool:
        """
        Create a custom role with specific permissions
        """
        try:
            from models.custom_role import CustomRole
            
            # Check if creator can create roles
            creator = self.db.query(User).filter(User.id == created_by).first()
            if not creator or creator.role not in ["admin", "super_admin"]:
                logger.error(f"User {created_by} cannot create custom roles")
                return False
            
            # Check if role already exists
            existing_role = self.db.query(CustomRole).filter(
                CustomRole.name == role_name
            ).first()
            
            if existing_role:
                logger.error(f"Custom role {role_name} already exists")
                return False
            
            # Create custom role
            custom_role = CustomRole(
                name=role_name,
                permissions=permissions,
                created_by=created_by,
                created_at=datetime.utcnow()
            )
            
            self.db.add(custom_role)
            self.db.commit()
            
            logger.info(f"Custom role {role_name} created by {created_by}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to create custom role: {str(e)}")
            self.db.rollback()
            return False
    
    def get_audit_log(self, user_id: Optional[str] = None, action: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get audit log for role and permission changes
        """
        try:
            from models.audit_log import AuditLog
            
            query = self.db.query(AuditLog).filter(
                AuditLog.action.in_(["role_assigned", "role_revoked", "permission_granted", "permission_denied"])
            )
            
            if user_id:
                query = query.filter(AuditLog.user_id == user_id)
            
            if action:
                query = query.filter(AuditLog.action == action)
            
            logs = query.order_by(AuditLog.created_at.desc()).limit(100).all()
            
            return [
                {
                    "id": log.id,
                    "user_id": log.user_id,
                    "action": log.action,
                    "details": log.details,
                    "ip_address": log.ip_address,
                    "user_agent": log.user_agent,
                    "created_at": log.created_at
                }
                for log in logs
            ]
            
        except Exception as e:
            logger.error(f"Failed to get audit log: {str(e)}")
            return []

# Global RBAC manager instance
rbac_manager = None

def get_rbac_manager(db: Session) -> RBACService:
    """
    Get RBAC manager instance
    """
    global rbac_manager
    if rbac_manager is None:
        rbac_manager = RBACService(db)
    return rbac_manager

File: frontend/lib/types/rbac.ts
/**
 * RBAC (Role-Based Access Control) Type Definitions
 *
 * Purpose: Type definitions for role-based access control, permissions, and authorization
 * in Buffr Host. Ensures secure access control across all system resources.
 * Location: lib/types/rbac.ts
 * Usage: Shared across all API routes, middleware, components, and services for authorization checks
 *
 * @module RBAC Types
 * @author Buffr Host Development Team
 * @version 1.0.0
 *
 * @description
 * This module defines the complete permission system for Buffr Host, including all available
 * permissions, roles, and access control contexts. All permissions follow the format:
 * 'resource:action' (e.g., 'bookings:read', 'customers:write').
 */

/**
 * Permission Enumeration
 *
 * Defines all available permissions in the Buffr Host system. Permissions are organized
 * by resource type and action level.
 *
 * @enum Permission
 * @property {string} USERS_READ - Read user information
 * @property {string} USERS_WRITE - Create/update users
 * @property {string} USERS_DELETE - Delete users
 * @property {string} USERS_MANAGE - Full user management access
 * @property {string} ROLES_READ - Read role information
 * @property {string} ROLES_WRITE - Create/update roles
 * @property {string} ROLES_DELETE - Delete roles
 * @property {string} ROLES_MANAGE - Full role management access
 * @property {string} PERMISSIONS_READ - Read permission information
 * @property {string} PERMISSIONS_WRITE - Create/update permissions
 * @property {string} PERMISSIONS_MANAGE - Full permission management access
 * @property {string} TENANTS_READ - Read tenant information
 * @property {string} TENANTS_WRITE - Create/update tenants
 * @property {string} TENANTS_DELETE - Delete tenants
 * @property {string} TENANTS_MANAGE - Full tenant management access
 * @property {string} PROPERTIES_READ - Read property information
 * @property {string} PROPERTIES_WRITE - Create/update properties
 * @property {string} PROPERTIES_DELETE - Delete properties
 * @property {string} PROPERTIES_MANAGE - Full property management access
 * @property {string} HOTEL_CONFIG_READ - Read hotel configuration
 * @property {string} HOTEL_CONFIG_WRITE - Create/update hotel configuration
 * @property {string} HOTEL_CONFIG_DELETE - Delete hotel configuration
 * @property {string} HOTEL_CONFIG_MANAGE - Full hotel configuration management
 * @property {string} BOOKINGS_READ - Read booking information
 * @property {string} BOOKINGS_WRITE - Create/update bookings
 * @property {string} BOOKINGS_DELETE - Delete bookings
 * @property {string} BOOKINGS_MANAGE - Full booking management access
 * @property {string} FINANCIAL_READ - Read financial information
 * @property {string} FINANCIAL_WRITE - Create/update financial records
 * @property {string} FINANCIAL_DELETE - Delete financial records
 * @property {string} FINANCIAL_MANAGE - Full financial management access
 * @property {string} SETTINGS_READ - Read system settings
 * @property {string} SETTINGS_WRITE - Update system settings
 * @property {string} SETTINGS_MANAGE - Full settings management access
 * @property {string} ANALYTICS_READ - Read analytics data
 * @property {string} ANALYTICS_WRITE - Create analytics reports
 * @property {string} ANALYTICS_MANAGE - Full analytics management access
 * @property {string} CUSTOMERS_READ - Read customer information
 * @property {string} CUSTOMERS_WRITE - Create/update customers
 * @property {string} CUSTOMERS_DELETE - Delete customers
 * @property {string} CUSTOMERS_MANAGE - Full customer management access
 * @property {string} STAFF_READ - Read staff information
 * @property {string} STAFF_WRITE - Create/update staff
 * @property {string} STAFF_DELETE - Delete staff
 * @property {string} STAFF_MANAGE - Full staff management access
 * @property {string} MENU_READ - Read menu information
 * @property {string} MENU_WRITE - Create/update menu items
 * @property {string} MENU_DELETE - Delete menu items
 * @property {string} MENU_MANAGE - Full menu management access
 * @property {string} ORDERS_READ - Read order information
 * @property {string} ORDERS_WRITE - Create/update orders
 * @property {string} ORDERS_DELETE - Delete orders
 * @property {string} ORDERS_MANAGE - Full order management access
 * @property {string} PAYMENTS_READ - Read payment information
 * @property {string} PAYMENTS_WRITE - Process payments
 * @property {string} PAYMENTS_DELETE - Delete payment records
 * @property {string} PAYMENTS_MANAGE - Full payment management access
 * @property {string} CMS_READ - Read CMS content
 * @property {string} CMS_WRITE - Create/update CMS content
 * @property {string} CMS_DELETE - Delete CMS content
 * @property {string} CMS_MANAGE - Full CMS management access
 * @property {string} LOYALTY_READ - Read loyalty program information
 * @property {string} LOYALTY_WRITE - Manage loyalty points
 * @property {string} LOYALTY_DELETE - Delete loyalty records
 * @property {string} LOYALTY_MANAGE - Full loyalty program management
 * @property {string} INVENTORY_READ - Read inventory information
 * @property {string} INVENTORY_WRITE - Update inventory levels
 * @property {string} INVENTORY_DELETE - Delete inventory records
 * @property {string} INVENTORY_MANAGE - Full inventory management access
 * @property {string} CALENDAR_READ - Read calendar information
 * @property {string} CALENDAR_WRITE - Create/update calendar events
 * @property {string} CALENDAR_DELETE - Delete calendar events
 * @property {string} CALENDAR_MANAGE - Full calendar management access
 * @property {string} SPA_READ - Read spa information
 * @property {string} SPA_WRITE - Create/update spa services
 * @property {string} SPA_DELETE - Delete spa services
 * @property {string} SPA_MANAGE - Full spa management access
 * @property {string} CONFERENCE_READ - Read conference room information
 * @property {string} CONFERENCE_WRITE - Create/update conference bookings
 * @property {string} CONFERENCE_DELETE - Delete conference bookings
 * @property {string} CONFERENCE_MANAGE - Full conference management access
 * @property {string} TRANSPORTATION_READ - Read transportation information
 * @property {string} TRANSPORTATION_WRITE - Create/update transportation bookings
 * @property {string} TRANSPORTATION_DELETE - Delete transportation bookings
 * @property {string} TRANSPORTATION_MANAGE - Full transportation management access
 * @property {string} AI_READ - Read AI/ML information
 * @property {string} AI_WRITE - Manage AI configurations
 * @property {string} AI_MANAGE - Full AI/ML management access
 * @property {string} REPORTS_READ - Read reports
 * @property {string} REPORTS_WRITE - Generate reports
 * @property {string} REPORTS_MANAGE - Full report management access
 * @property {string} NOTIFICATIONS_READ - Read notifications
 * @property {string} NOTIFICATIONS_WRITE - Send notifications
 * @property {string} NOTIFICATIONS_MANAGE - Full notification management access
 * @property {string} EMAIL_READ - Read email information
 * @property {string} EMAIL_WRITE - Send emails
 * @property {string} EMAIL_MANAGE - Full email management access
 * @property {string} FILES_READ - Read file information
 * @property {string} FILES_WRITE - Upload files
 * @property {string} FILES_DELETE - Delete files
 * @property {string} FILES_MANAGE - Full file management access
 *
 * @security Permissions are checked at the API route level and component level
 * @see {@link PermissionScope} for permission scope levels
 */
/**
 * Rbac Type Definitions for Buffr Host Hospitality Platform
 * @fileoverview Rbac type definitions for role-based access control and permission management
 * @location buffr-host/lib/types/rbac.ts
 * @purpose rbac type definitions for role-based access control and permission management
 * @modularity Centralized type definitions providing type safety across the entire Buffr Host application
 * @api_integration REST API endpoints, HTTP request/response handling
 * @security Type-safe security definitions for authentication, authorization, and data protection
 * @ai_integration Machine learning and AI service type definitions for predictive analytics
 * @authentication User authentication and session management type definitions
 * @typescript_strict Strict TypeScript type safety ensuring compile-time error prevention
 * @type_safety Comprehensive type coverage preventing runtime errors and improving developer experience
 * @scalability Type definitions supporting horizontal scaling and multi-tenant architecture
 * @maintainability Self-documenting types enabling easier code maintenance and refactoring
 * @testing Type-driven development supporting comprehensive test coverage
 *
 * Type Definitions Summary:
 * - 5 Interfaces: UserPermission, UserRoleAssignment, RolePermission...
 * - 4 Enums: Permission, Role, UserRole...
 * - Total: 9 type definitions
 *
 * Usage and Integration:
 * - Frontend Components: Type-safe props and state management
 * - API Routes: Request/response type validation
 * - Database Services: Schema-aligned data operations
 * - Business Logic: Domain-specific type constraints
 * - Testing: Type-driven test case generation
 *
 * @example
 * // Import type definitions
 * import type { Permission, Role, UserRole... } from './rbac';
 *
 * // Usage in component
 * interface ComponentProps {
 *   data: UserPermission;
 *   onAction: (event: EventType) => void;
 * }
 *
 * @example
 * // Database service usage
 * const userService = {
 *   async getUser(id: string): Promise<UserPermission> {
 *     // Type-safe database operations
 *     return await db.query('SELECT * FROM users WHERE id = $1', [id]);
 *   }
 * };
 *
 * Exported Types:
 * @typedef {Enum} Permission
 * @typedef {Enum} Role
 * @typedef {Enum} UserRole
 * @typedef {Enum} PermissionScope
 * @typedef {Interface} UserPermission
 * @typedef {Interface} UserRoleAssignment
 * @typedef {Interface} RolePermission
 * @typedef {Interface} PermissionCheck
 * @typedef {Interface} RBACContext
 *
 * @returns {Object} Type definitions module with all exported types and interfaces
 * @type_safety Ensures 100% type coverage across the Buffr Host application
 * @documentation Comprehensive JSDoc documentation for developer experience
 * @maintainability Type-driven development enabling safe refactoring and evolution
 */

export enum Permission {
  // User management
  USERS_READ = 'users:read',
  USERS_WRITE = 'users:write',
  USERS_DELETE = 'users:delete',
  USERS_MANAGE = 'users:manage',

  // Role management
  ROLES_READ = 'roles:read',
  ROLES_WRITE = 'roles:write',
  ROLES_DELETE = 'roles:delete',
  ROLES_MANAGE = 'roles:manage',

  // Permission management
  PERMISSIONS_READ = 'permissions:read',
  PERMISSIONS_WRITE = 'permissions:write',
  PERMISSIONS_MANAGE = 'permissions:manage',

  // Tenant management
  TENANTS_READ = 'tenants:read',
  TENANTS_WRITE = 'tenants:write',
  TENANTS_DELETE = 'tenants:delete',
  TENANTS_MANAGE = 'tenants:manage',

  // Property management
  PROPERTIES_READ = 'properties:read',
  PROPERTIES_WRITE = 'properties:write',
  PROPERTIES_DELETE = 'properties:delete',
  PROPERTIES_MANAGE = 'properties:manage',

  // Hotel configuration
  HOTEL_CONFIG_READ = 'hotel_configuration:read',
  HOTEL_CONFIG_WRITE = 'hotel_configuration:write',
  HOTEL_CONFIG_DELETE = 'hotel_configuration:delete',
  HOTEL_CONFIG_MANAGE = 'hotel_configuration:manage',

  // Booking management
  BOOKINGS_READ = 'bookings:read',
  BOOKINGS_WRITE = 'bookings:write',
  BOOKINGS_DELETE = 'bookings:delete',
  BOOKINGS_MANAGE = 'bookings:manage',

  // Financial management
  FINANCIAL_READ = 'financial:read',
  FINANCIAL_WRITE = 'financial:write',
  FINANCIAL_DELETE = 'financial:delete',
  FINANCIAL_MANAGE = 'financial:manage',

  // System management
  SETTINGS_READ = 'settings:read',
  SETTINGS_WRITE = 'settings:write',
  SETTINGS_MANAGE = 'settings:manage',

  // Analytics
  ANALYTICS_READ = 'analytics:read',
  ANALYTICS_WRITE = 'analytics:write',
  ANALYTICS_MANAGE = 'analytics:manage',

  // Customer management
  CUSTOMERS_READ = 'customers:read',
  CUSTOMERS_WRITE = 'customers:write',
  CUSTOMERS_DELETE = 'customers:delete',
  CUSTOMERS_MANAGE = 'customers:manage',

  // Staff management
  STAFF_READ = 'staff:read',
  STAFF_WRITE = 'staff:write',
  STAFF_DELETE = 'staff:delete',
  STAFF_MANAGE = 'staff:manage',

  // Menu management
  MENU_READ = 'menu:read',
  MENU_WRITE = 'menu:write',
  MENU_DELETE = 'menu:delete',
  MENU_MANAGE = 'menu:manage',

  // Order management
  ORDERS_READ = 'orders:read',
  ORDERS_WRITE = 'orders:write',
  ORDERS_DELETE = 'orders:delete',
  ORDERS_MANAGE = 'orders:manage',

  // Payment management
  PAYMENTS_READ = 'payments:read',
  PAYMENTS_WRITE = 'payments:write',
  PAYMENTS_DELETE = 'payments:delete',
  PAYMENTS_MANAGE = 'payments:manage',

  // CMS management
  CMS_READ = 'cms:read',
  CMS_WRITE = 'cms:write',
  CMS_DELETE = 'cms:delete',
  CMS_MANAGE = 'cms:manage',

  // Loyalty management
  LOYALTY_READ = 'loyalty:read',
  LOYALTY_WRITE = 'loyalty:write',
  LOYALTY_DELETE = 'loyalty:delete',
  LOYALTY_MANAGE = 'loyalty:manage',

  // Inventory management
  INVENTORY_READ = 'inventory:read',
  INVENTORY_WRITE = 'inventory:write',
  INVENTORY_DELETE = 'inventory:delete',
  INVENTORY_MANAGE = 'inventory:manage',

  // Calendar management
  CALENDAR_READ = 'calendar:read',
  CALENDAR_WRITE = 'calendar:write',
  CALENDAR_DELETE = 'calendar:delete',
  CALENDAR_MANAGE = 'calendar:manage',

  // Spa management
  SPA_READ = 'spa:read',
  SPA_WRITE = 'spa:write',
  SPA_DELETE = 'spa:delete',
  SPA_MANAGE = 'spa:manage',

  // Conference management
  CONFERENCE_READ = 'conference:read',
  CONFERENCE_WRITE = 'conference:write',
  CONFERENCE_DELETE = 'conference:delete',
  CONFERENCE_MANAGE = 'conference:manage',

  // Transportation management
  TRANSPORTATION_READ = 'transportation:read',
  TRANSPORTATION_WRITE = 'transportation:write',
  TRANSPORTATION_DELETE = 'transportation:delete',
  TRANSPORTATION_MANAGE = 'transportation:manage',

  // AI/ML management
  AI_READ = 'ai:read',
  AI_WRITE = 'ai:write',
  AI_MANAGE = 'ai:manage',

  // Reports management
  REPORTS_READ = 'reports:read',
  REPORTS_WRITE = 'reports:write',
  REPORTS_MANAGE = 'reports:manage',

  // Notifications management
  NOTIFICATIONS_READ = 'notifications:read',
  NOTIFICATIONS_WRITE = 'notifications:write',
  NOTIFICATIONS_MANAGE = 'notifications:manage',

  // Email management
  EMAIL_READ = 'email:read',
  EMAIL_WRITE = 'email:write',
  EMAIL_MANAGE = 'email:manage',

  // File management
  FILES_READ = 'files:read',
  FILES_WRITE = 'files:write',
  FILES_DELETE = 'files:delete',
  FILES_MANAGE = 'files:manage',
}

/**
 * Role Enumeration
 *
 * Defines the predefined roles in the Buffr Host system. Roles are assigned to users
 * and determine their default permissions.
 *
 * @enum Role
 * @property {string} SUPER_ADMIN - Platform super administrator with full system access
 * @property {string} ADMIN - Tenant administrator with full tenant access
 * @property {string} MANAGER - Property manager with management-level access
 * @property {string} STAFF - Staff member with operational access
 * @property {string} GUEST - Guest user with read-only access
 */
export enum Role {
  SUPER_ADMIN = 'super_admin',
  ADMIN = 'admin',
  MANAGER = 'manager',
  STAFF = 'staff',
  GUEST = 'guest',
}

export enum UserRole {
  SUPER_ADMIN = 'super_admin',
  ADMIN = 'admin',
  MANAGER = 'manager',
  STAFF = 'staff',
  GUEST = 'guest',
}

export enum PermissionScope {
  GLOBAL = 'global', // System-wide access
  TENANT = 'tenant', // Tenant-specific access
  PROPERTY = 'property', // Property-specific access
  USER = 'user', // User-specific access
}

export interface UserPermission {
  permission: Permission;
  granted: boolean;
  source: string; // role, custom, etc.
}

export interface UserRoleAssignment {
  user_id: string;
  role: UserRole;
  assigned_by: string;
  assigned_at: Date;
}

export interface RolePermission {
  role: Role;
  permissions: Permission[];
  scope: PermissionScope;
}

export interface PermissionCheck {
  user_id: string;
  permission: Permission;
  resource_id?: string;
  scope?: PermissionScope;
}

/**
 * RBAC Context Interface
 *
 * Represents the authorization context for a user, including their role, permissions,
 * and tenant/property associations. Used throughout the application for access control checks.
 *
 * @interface RBACContext
 * @property {string} user_id - User ID for authorization checks
 * @property {UserRole} role - User's role in the system
 * @property {Permission[]} permissions - Array of permissions granted to the user
 * @property {string} [tenant_id] - Tenant ID for tenant-scoped operations
 * @property {string} [property_id] - Property ID for property-scoped operations
 *
 * @security This context must be validated on every API request
 * @example
 * const context: RBACContext = {
 *   user_id: 'user_123',
 *   role: UserRole.MANAGER,
 *   permissions: [Permission.BOOKINGS_READ, Permission.BOOKINGS_WRITE],
 *   tenant_id: 'tenant_abc',
 *   property_id: 'prop_xyz'
 * };
 */
export interface RBACContext {
  user_id: string;
  role: UserRole;
  permissions: Permission[];
  tenant_id?: string;
  property_id?: string;
}
